File: 1043. Partition Array for Maximum Sum.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {

        int[] res = new int[arr.length];

        for(int i = 0; i < arr.length; i++){
            int max = Integer.MIN_VALUE;
            int total = Integer.MIN_VALUE;
            for(int j = i; j >= 0 && i-j < k; j--){
                max = Math.max(arr[j], max);
                int me = ((1+i)- j)*max + (j > 0 ? res[j-1] : 0);
                total = Math.max(total, me);
            }
            res[i] = total;
        }

        //for(int num : res) System.out.print(num + "\t");

        return res[res.length-1];
        
    }
}

File: 1043. Partition Array for Maximum Sum.cpp
/*
Time: O(n*k)
Space: O(n)
*/

class Solution {
public:
    vector<int> v, dp;
    int n,k;

    int call(int pos){
      if(pos==n)  return 0;
      if(dp[pos]!=-1)  return dp[pos];
      
      int maxi = 0;
      int ans = 0;

      //options: take this element in the subarray (check if it's the maximum for this subarray)
      //opt1 = continue till subarray size <=k
      //opt2 = end this subarray and recusive call to build next subarray from the next element
      for(int i=pos; i<min(n,pos+k); i++){
        maxi = max(maxi, v[i]);
        int subarr_sum = (i-pos+1) * maxi;
        ans = max(ans, subarr_sum + call(i+1));
      }

      return dp[pos] = ans;
    }


    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        v = arr;
        n = arr.size();
        dp = vector(n,-1);
        this->k = k;

        return call(0);
    }
};
